### PR TITLE
Add UpsertMemo API

### PR DESCRIPTION
**Pull request by rodrigozhou**
## What was changed

Adding an API for `UpsertMemo`. It merges with current memo and creates a command `ModifyWorkflowProperties` with the updated memo.
This change depends on https://github.com/temporalio/api/pull/218.

## Why?

Feature request: https://github.com/temporalio/sdk-features/issues/119

## Checklist

1. How was this tested:
   - Created new tests.

### CODE REVIEW

File: internal/internal_event_handlers.go
Comment (cretz) [inline internal/internal_event_handlers.go:387]:
```suggestion
func (wc *workflowEnvironmentImpl) UpsertMemo(memoMap map[string]interface{}) error {
```

Here and elsewhere

File: internal/internal_event_handlers.go
Comment (cretz) [inline internal/internal_event_handlers.go:407]: Could also just `len(current.GetFields()) == 0`

File: test/integration_test.go
Comment (cretz) [inline test/integration_test.go:1456]: Can you provide a full integration test here? I think a good test would be to start a workflow with some memo, assert that memo, then upsert that memo (and set a query saying you've done so), then describe the workflow via the client and confirm the memo was upserted.

If this is not available in the server yet for an integration test, I don't think we should merge the SDK functionality until it is.

**Comment by bergundy** [comment]
Didn't look at the code but something I wish we'd done with search attributes in TS is delay sending the command until the end of an "activation", so if a user calls `UpsertMemo({ a: 1 }); UpsertMemo({ b: 2 }); UpsertMemo({ a: 2 })` the SDK would only issue 1 `UpsertMemo({ a: 2, b: 2})` command.

**Comment by bergundy** [comment]
@rodrigozhou note that I moved the original issue to sdk-features and marked this PR as not closing for that issue, we need to implement in all SDKs.

File: internal/internal_event_handlers.go
Comment (rodrigozhou) [reply internal/internal_event_handlers.go:402]: I'm naming it generic because so it can be extended later when new attributes are added.
I can make it specific, but I suppose new attributes would require creating individual functions to update each of them.

Comment (cretz) [reply internal/internal_event_handlers.go:402]: The parameter is very specific though, so it's obviously not capable of future attributes without altering the signature anyways. So might as well not have a generic name w/ one specific param for the one specific use case.

File: internal/internal_event_handlers.go
Comment (rodrigozhou) [reply internal/internal_event_handlers.go:407]: In theory, the new memo cannot be nil or an empty map. See `validateAndSerializeMemo` which is executed before merging the memo.

**Comment by rodrigozhou** [comment]
The CI failure is because the CI used the docker-compose repo to start the temporal server, which doesn't include `UpsertMemo` yet. Locally, all tests passes.

File: test/integration_test.go
Comment (cretz) [inline test/integration_test.go:2522]: My approval of the PR is pending @bergundy and @yiminc and others coming to a conclusion on whether we want a capability boolean so users can check whether their server version supports this.

File: workflow/workflow.go
Comment (cretz) [inline workflow/workflow.go:473]: Every exported function should have Godoc, especially this one :-)

File: test/integration_test.go
Comment (rodrigozhou) [reply test/integration_test.go]: Updated to check capability.

File: test/integration_test.go
Comment (cretz) [inline test/integration_test.go:2565]: I think a simple `DescribeWorkflowExecution` is much better here than listing. But I'm gonna mark this approved anyways, change only if you want.

**Review by cretz — APPROVED** [review]

File: workflow/workflow.go
Comment (cretz) [inline workflow/workflow.go:497]: We should probably make it clear not all servers support this (and maybe say which server version does)

File: internal/internal_task_handlers.go
Comment (yiminc) [inline internal/internal_task_handlers.go:1454]: StrictMode was there for historical reason. But I don't think they are needed anymore. We probably should remove strictMode. @cretz 
(No change requested for this PR.)

**Pull request merged**
